reinforcement_learning/include/model_mgmt.h

```diff
@@ -26,20 +26,21 @@ namespace reinforcement_learning { namespace model_management {
         model_data();
         ~model_data();
 
-        model_data(model_data const& other) = delete;
-        model_data& operator=(model_data const& other) = delete;
+        model_data(model_data const& other);
+        model_data& operator=(model_data const& other);
 
         model_data(model_data&& other) noexcept
           : _data(other._data),
             _data_sz(other._data_sz),
             _refresh_count(other._refresh_count) {}
 
         model_data& operator=(model_data&& other) noexcept {
-          if (this == &other)
-            return *this;
-          _data = other._data;
-          _data_sz = other._data_sz;
-          _refresh_count = other._refresh_count;
+          if (this != &other) {
+            _data = other._data;
```

Comment:
Should we not swap content here to be correct?

Reply:
We don't want to set the state of other to what the current object holds. Unless we explicitly set ourselves to null/default values and then do a swap

Reply:
Will this cause a memory leak if a temporary is assigned to a named model_data?


---
In reply to: [222718213](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1605#discussion_r222718213) [](ancestors = 222718213)

Reply:
This is a good point. We need to free any existing memory before we accept from the other object. We also need to be careful to set the other object to null so that when it is destroyed it doesnt clear the memory. I'll fix this up

Reply:
doing the swap should take care of that for you.

---
In reply to: [222725336](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1605#discussion_r222725336) [](ancestors = 222725336)

Reply:
Not if the current object already has its members populated though right?